src/vivarium/component.py

```diff
@@ -142,67 +147,63 @@ def set_population_view(self, builder: "Builder") -> None:
+                requires_streams=self.initialization_columns_required,
+            )
 
     def register_time_step_prepare_listener(self, builder: "Builder") -> None:
         """
         Registers a time-step prepare listener if this component has defined
         one.
         """
-        builder.event.register_listener(
-            "time_step__prepare",
-            self.on_time_step_prepare,
-            self.time_step_prepare_priority,
-        )
+        if self.on_time_step_prepare is not None:
+            builder.event.register_listener(
+                "time_step__prepare",
+                self.on_time_step_prepare,
+                self.time_step_prepare_priority,
+            )
 
     def register_time_step_listener(self, builder: "Builder") -> None:
         """Registers a time-step listener if this component has defined one."""
-        builder.event.register_listener(
-            "time_step",
-            self.on_time_step,
-            self.time_step_priority,
-        )
+        if self.on_time_step is not None:
+            builder.event.register_listener(
+                "time_step",
+                self.on_time_step,
+                self.time_step_priority,
+            )
 
     def register_time_step_cleanup_listener(self, builder: "Builder") -> None:
         """
         Registers a time-step cleanup listener if this component has defined
         one.
         """
-        builder.event.register_listener(
-            "time_step__cleanup",
-            self.on_time_step_cleanup,
-            self.time_step_cleanup_priority,
-        )
+        if self.on_time_step_cleanup is not None:
+            builder.event.register_listener(
+                "time_step__cleanup",
+                self.on_time_step_cleanup,
+                self.time_step_cleanup_priority,
+            )
 
     def register_collect_metrics_listener(self, builder: "Builder") -> None:
         """
         Registers a collect metrics listener if this component has defined one.
         """
-        builder.event.register_listener(
-            "collect_metrics",
-            self.on_collect_metrics,
-            self.collect_metrics_priority,
-        )
+        if self.on_collect_metrics is not None:
+            builder.event.register_listener(
+                "collect_metrics",
+                self.on_collect_metrics,
+                self.collect_metrics_priority,
+            )
 
     ########################
     # Event-driven methods #
     ########################
 
-    def on_initialize_simulants(self, pop_data: "SimulantData") -> None:
-        pass
-
-    def on_time_step_prepare(self, event: Event) -> None:
-        pass
-
-    def on_time_step(self, event: Event) -> None:
-        pass
-
-    def on_time_step_cleanup(self, event: Event) -> None:
-        pass
-
-    def on_collect_metrics(self, event: Event) -> None:
-        pass
+    on_initialize_simulants: Optional[Callable[["SimulantData"], None]] = None
```

Comment:
This is just defining `on_initialize_simulants` as an attribute of `Component` and setting its value to `None`. If a class subclasses `Component` and defines a method called `on_initialize_simulants` it will override this. In the method `register_simulant_initializer` you can see that we check if this attribute is `None` and if it is, we do nothing. That is essentially a check as to whether this method has been defined in the subclass.